File: BoyiaApp/ios/app/app/Model/BoyiaLoginModel.swift
//
//  BoyiaLoginModel.swift
//  app
//
//  Created by yanbo on 2022/5/24.
//

import Foundation

class BoyiaLoginModel : ObservableObject {
    @Published var isLogin = false
    
    init() {
        BoyiaLog.d("token=\(String(describing: BoyiaLoginInfo.shared.token))")
        isLogin = BoyiaLoginInfo.shared.token != nil
    }
    
    func login(name: String, password: String, onComplete: ((Bool) -> Void)? = nil) {
        if name.isEmpty || password.isEmpty {
            BoyiaLog.d("name and password is empty")
            return
        }
        BoyiaLog.d("name=\(name)&pwd=\(password)")
        //var headers = NSMutableDictionary()
        //headers.setObject("application/x-www-form-urlencoded", forKey: NSString("Content-Type"))
        var headers = [AnyHashable: Any]()
        headers["Content-Type"] = "application/x-www-form-urlencoded"
        headers["User-Token"] = "none"
        
        let data = "name=\(name)&pwd=\(password)"
        
        HttpUtil.post(url: HttpUtil.HttpConstants.LOGIN_URL, data: data, headers: headers, cb: { (data: BoyiaUserData) in
            DispatchQueue.main.async {
                BoyiaLog.d("BoyiaLoginModel login = \(data.retCode)")
                if (data.retCode == 200) {
                    BoyiaLoginInfo.setLoginInfo(info: data.data, token: data.userToken)
                    self.isLogin = true
                    if onComplete != nil {
                        onComplete!(true)
                    }
                } else {
                    if onComplete != nil {
                        onComplete!(false)
                    }
                }
            }
        })
    }
    
    func logout(onComplete: ((Bool) -> Void)? = nil) {
        var headers = [AnyHashable: Any]()
        headers["Content-Type"] = "application/x-www-form-urlencoded"
        headers["User-Token"] = BoyiaLoginInfo.shared.token
        HttpUtil.post(url: HttpUtil.HttpConstants.LOGOUT_URL, headers: headers, cb: { (data: BoyiaBaseData) in
            DispatchQueue.main.async {
                BoyiaLog.d("BoyiaLoginModel logout = \(data.retCode)")
                // 就算失败本地也不再保存变量
                BoyiaLoginInfo.logout()
                self.isLogin = false
            }
        })
    }
}

File: BoyiaApp/android/app/src/main/kotlin/com/boyia/app/shell/model/BoyiaLoginModel.kt
package com.boyia.app.shell.model

import com.boyia.app.common.json.BoyiaJson
import com.boyia.app.common.utils.BoyiaLog
import com.boyia.app.common.utils.BoyiaUtils
import com.boyia.app.loader.http.HTTPFactory
import com.boyia.app.loader.mue.MainScheduler
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.coroutines.suspendCoroutine

object BoyiaLoginModel {
    const val TAG = "BoyiaLoginModel"

    /**
     * 使用协程做登录请求
     */
    suspend fun suspendLogin(name: String?, password: String?) : BoyiaUserData {
        return withContext(Dispatchers.IO) {
            /**
             * 处理回调
             */
            suspendCoroutine {
                BoyiaModelUtil.request(
                        BoyiaModelUtil.LOGIN_URL,
                        object: BoyiaModelUtil.ModelDataCallback<BoyiaUserData> {
                            override fun onLoadData(data: BoyiaUserData) {
                                BoyiaLog.d(TAG, "retCode = " + data.retCode)
                                if (data.retCode != 200) {
                                    return
                                }

                                it.resume(data)
                            }
                        },
                        method = HTTPFactory.HTTP_POST_METHOD,
                        headers = mapOf(HTTPFactory.HeaderKeys.CONTENT_TYPE to HTTPFactory.HeaderValues.FORM, "User-Token" to "none"),
                        data = "name=${name}&pwd=${password}"
                )
            }
        }
    }

    fun login(name: String?, password: String?, callback: (info: BoyiaUserInfo) -> Unit) {
        if (BoyiaUtils.isTextEmpty(name)
                || BoyiaUtils.isTextEmpty(password)
        ) {
            BoyiaLog.d(TAG, "name or passsword is emprt")
            return
        }

        BoyiaModelUtil.request(
                BoyiaModelUtil.LOGIN_URL,
                object: BoyiaModelUtil.ModelDataCallback<BoyiaUserData> {
                    override fun onLoadData(data: BoyiaUserData) {
                        BoyiaLog.d(TAG, "retCode = " + data.retCode)
                        if (data.retCode != 200) {
                            return
                        }

                        MainScheduler.mainScheduler().sendJob {
                            BoyiaLoginInfo.instance().token = data.userToken
                            BoyiaLoginInfo.instance().user = data.data

                            data.data?.let { callback(it) }
                        }
                    }
                },
                method = HTTPFactory.HTTP_POST_METHOD,
                headers = mapOf(HTTPFactory.HeaderKeys.CONTENT_TYPE to HTTPFactory.HeaderValues.FORM, "User-Token" to "none"),
                data = "name=${name}&pwd=${password}"
        )
    }

    fun update() {
        val info =  BoyiaJson.toJson(BoyiaLoginInfo.instance().user)
        BoyiaModelUtil.request<BoyiaBaseData>(
                BoyiaModelUtil.UPDATE_USER_URL,
                method = HTTPFactory.HTTP_POST_METHOD,
                headers = mapOf(
                        "User-Token" to (BoyiaLoginInfo.instance().token ?: "none"),
                        HTTPFactory.HeaderKeys.CONTENT_TYPE to HTTPFactory.HeaderValues.JSON
                ),
                data = info
        )
    }

    fun logout() {
        BoyiaModelUtil.request<BoyiaBaseData>(
                BoyiaModelUtil.LOGOUT_URL,
                method = HTTPFactory.HTTP_POST_METHOD,
                headers = mapOf("User-Token" to (BoyiaLoginInfo.instance().token ?: "none"))
        )
    }
}